File: d05/d05.py
from functools import reduce

stack, moves = open(0).read().split('\n\n')
stack = [[]] + [''.join(l).strip()[::-1] for l in zip(*[l[1::4] for l in stack.split('\n')[:-1]])]
moves = [[*map(int, m[1::2])] for m in map(str.split, moves.split('\n')[:-1])]

def f(acc, move):
    st, dx = acc
    cnt, p_from, p_to = move
    st[p_to] += st[p_from][-cnt:][::dx]
    st[p_from] = st[p_from][:-cnt]
    return st, dx

star = lambda dx, st: ''.join(c[-1] for c in reduce(f, moves, (st, dx))[0][1:])
print(star(-1, stack.copy()), star(1, stack.copy()))

File: d05/d05.exs
parse_table = fn table ->
  table
  |> String.split("\n") 
  |> Enum.drop(-1)
  |> Enum.map(& 
    &1
    |> to_charlist()
    |> Enum.drop(1)
    |> Enum.take_every(4)
  )
  |> Enum.zip()
  |> Enum.map(fn pile ->
    pile
    |> Tuple.to_list()
    |> Enum.drop_while(& &1 == ?\s)
  end)
  |> then(& ['' | &1])
end

parse_commands = fn commands ->
  commands
  |> String.split("\n")
  |> Enum.drop(-1)
  |> Enum.map(fn command ->
    Regex.run(~r{move (\d+) from (\d+) to (\d+)}, command) 
    |> tl() 
    |> Enum.map(&String.to_integer/1)
  end)
end

solve = fn [reverse?: reverse?] ->
  {table, commands} = File.read!("input.txt")
    |> String.split("\n\n")
    |> then(fn [table, commands] -> {parse_table.(table), parse_commands.(commands)} end)
  
  Enum.reduce(commands, table, fn [cnt, from, to], acc ->
    {move, rest} = acc
      |> Enum.at(from)
      |> Enum.split(cnt)
  
    acc
      |> List.update_at(to, & 
        case reverse? do
          true -> Enum.reverse(move)
          false -> move
        end
        ++ &1
      )
      |> List.replace_at(from, rest)
  end)
  |> tl()
  |> Enum.map(&hd/1)
  |> to_string()
end

IO.inspect(solve.(reverse?: true), label: "Star 1")
IO.inspect(solve.(reverse?: false), label: "Star 2")
